File: alexandria-markup-server/src/main/java/nl/knaw/huygens/alexandria/dropwizard/cli/commands/ValidateCommand.java
package nl.knaw.huygens.alexandria.dropwizard.cli.commands;

/*
 * #%L
 * alexandria-markup-server
 * =======
 * Copyright (C) 2015 - 2019 Huygens ING (KNAW)
 * =======
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.google.common.base.Charsets;
import io.dropwizard.setup.Bootstrap;
import net.sourceforge.argparse4j.inf.Namespace;
import net.sourceforge.argparse4j.inf.Subparser;
import nl.knaw.huc.di.tag.schema.TAGMLSchemaFactory;
import nl.knaw.huc.di.tag.schema.TAGMLSchemaParseResult;
import nl.knaw.huc.di.tag.validate.TAGValidationResult;
import nl.knaw.huc.di.tag.validate.TAGValidator;
import nl.knaw.huygens.alexandria.dropwizard.cli.CLIContext;
import nl.knaw.huygens.alexandria.storage.TAGDocument;
import nl.knaw.huygens.alexandria.storage.TAGStore;
import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;

import static java.util.stream.Collectors.joining;

public class ValidateCommand extends AlexandriaCommand {

  private static final String SCHEMA = "schema";

  public ValidateCommand() {
    super("schema-validate", "Validate a document against a TAG schema.");
  }

  @Override
  public void configure(Subparser subparser) {
    subparser
        .addArgument("-d", "--document") //
        .dest(DOCUMENT) //
        .type(String.class) //
        .required(true) //
        .help("The name of the document to validate.");
    subparser
        .addArgument("-s", "--schema") //
        .dest(SCHEMA)
        .type(String.class) //
        .required(true) //
        .help("The TAG schema file.");
  }

  @Override
  public void run(Bootstrap<?> bootstrap, Namespace namespace) throws IOException {
    checkAlexandriaIsInitialized();

    String docName = namespace.getString(DOCUMENT);
    Long docId = getIdForExistingDocument(docName);

    String schemaFile = namespace.getString(SCHEMA);
    Path filePath = workFilePath(schemaFile);
    File file = filePath.toFile();
    if (file.isFile()) {
      String schemaYAML = FileUtils.readFileToString(file, Charsets.UTF_8);
      final TAGMLSchemaParseResult schemaParseResult = TAGMLSchemaFactory.parseYAML(schemaYAML);

      CLIContext context = readContext();

      try (TAGStore store = getTAGStore()) {
        TAGValidationResult result =
            store.runInTransaction(
                () -> {
                  TAGDocument document = store.getDocument(docId);
                  return new TAGValidator(store).validate(document, schemaParseResult.schema);
                });

        System.out.println("Document " + docName + " is ");
        if (!result.isValid()) {
          System.out.println("  not valid:");
          System.out.println(
              result.errors.stream().map(e -> "  - error: " + e).collect(joining("\n")));
        } else {
          System.out.println("  valid");
        }
        System.out.println("according to the schema defined in " + schemaFile);
        System.out.println(
            result.warnings.stream().map(e -> "  - warning: " + e).collect(joining("\n")));
      }
    } else {
      System.err.printf("%s is not a file!%n", schemaFile);
    }
  }
}

File: alexandria-markup-server/src/main/kotlin/nl/knaw/huygens/alexandria/dropwizard/cli/commands/ValidateCommand.kt
package nl.knaw.huygens.alexandria.dropwizard.cli.commands

/*
* #%L
 * alexandria-markup-server
 * =======
 * Copyright (C) 2015 - 2020 Huygens ING (KNAW)
 * =======
 * Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  
 *       http://www.apache.org/licenses/LICENSE-2.0
 *  
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 * #L%
*/

import com.google.common.base.Charsets
import io.dropwizard.setup.Bootstrap
import net.sourceforge.argparse4j.inf.Namespace
import net.sourceforge.argparse4j.inf.Subparser
import nl.knaw.huc.di.tag.schema.TAGMLSchemaFactory.parseYAML
import nl.knaw.huc.di.tag.schema.TAGMLSchemaParseResult
import nl.knaw.huc.di.tag.validate.TAGValidator
import nl.knaw.huygens.alexandria.dropwizard.cli.AlexandriaCommandException
import nl.knaw.huygens.alexandria.storage.TAGDocument
import nl.knaw.huygens.alexandria.storage.TAGStore
import org.apache.commons.io.IOUtils
import java.io.IOException
import java.net.URL
import java.util.stream.Collectors

class ValidateCommand : AlexandriaCommand("schema-validate", "Validate a document against a TAG schema.") {
    override fun configure(subparser: Subparser) {
        subparser
                .addArgument("DOCUMENT")
                .metavar("<document>")
                .dest(DOCUMENT)
                .type(String::class.java)
                .required(true)
                .help(
                        "The name of the document to validate. It must have a valid URL to a valid schema file (in YAML) defined using '[!schema <schemaLocationURL>]' in the TAGML source file.")
        //    subparser
        //        .addArgument("-s", "--schema")
        //        .dest(SCHEMA)
        //        .type(String.class)
        //        .required(true)
        //        .help("The TAG schema file.");
    }

    @Throws(IOException::class)
    override fun run(bootstrap: Bootstrap<*>?, namespace: Namespace) {
        checkAlexandriaIsInitialized()
        val docName = namespace.getString(DOCUMENT)
        val docId = getIdForExistingDocument(docName)
        tagStore.use { store ->
            store.runInTransaction {
                val document = store.getDocument(docId)
                try {
                    val schemaLocation = document.schemaLocation
                    if (schemaLocation.isPresent) {
                        continueWithSchemaLocation(docName, store, document, schemaLocation.get())
                    } else {
                        throw AlexandriaCommandException(
                                """There was no schema location defined in $docName, please add
  [!schema <schemaLocationURL>]
to the tagml sourcefile.""")
                    }
                } catch (e: IOException) {
                    throw AlexandriaCommandException(
                            "The schema location in $docName is invalid.")
                }
            }
        }
    }

    @Throws(IOException::class)
    private fun continueWithSchemaLocation(
            docName: String, store: TAGStore, document: TAGDocument, url: URL) {
        var tmpSchemaLocationURL = url.toString()
        if (url.protocol == "file" && !tmpSchemaLocationURL.contains("file://")) {
            tmpSchemaLocationURL = (url.protocol
                    + "://"
                    + url.path) // because url.toString() somehow loses     '//', but only on
            // Windows?!
        }
        val schemaLocationURL = tmpSchemaLocationURL
        println("Parsing schema from $schemaLocationURL:")
        val schemaYAML = IOUtils.toString(url, Charsets.UTF_8)
        val schemaParseResult = parseYAML(schemaYAML)
        if (schemaParseResult.errors.isEmpty()) {
            continueWithValidSchema(docName, store, document, schemaParseResult, schemaLocationURL)
        } else {
            println("  errors:\n"
                    + schemaParseResult.errors.stream()
                    .map { e: String -> "  - " + e.replace("\\(StringReader\\)".toRegex(), schemaLocationURL) }
                    .collect(Collectors.joining("\n")))
        }
    }

    private fun continueWithValidSchema(
            docName: String,
            store: TAGStore,
            document: TAGDocument,
            schemaParseResult: TAGMLSchemaParseResult,
            schemaLocationURL: String) {
        println("  done\n")
        val result = TAGValidator(store).validate(document, schemaParseResult.schema)
        println("Document $docName is ")
        if (!result.isValid) {
            println("  not valid:")
            println(
                    result.errors.stream().map { e: String -> "  - error: $e" }.collect(Collectors.joining("\n")))
        } else {
            println("  valid")
        }
        println(
                result.warnings.stream().map { e: String -> "  - warning: $e" }.collect(Collectors.joining("\n")))
        println("according to the schema defined in $schemaLocationURL")
    }

    companion object {
        private const val SCHEMA = "schema"
    }
}
